### PR TITLE
Added functionality for symbolically computing DAE jacobians

### LINKED ISSUE #1482: `jac` and `sparse` are not implemented for DAEProblem
```julia
using OrdinaryDiffEq, ModelingToolkit
using BenchmarkTools
using Plots
using Sundials

function build_system(n)

	@parameters t # or would @variables t be better!?
	@variables Tgas[1:n](t) Tsol[1:n](t) cR[1:n](t) mflux(t)
	@parameters params[1:3]
	D = Differential(t)


    dp, Tin, h = params

    dx = h / n

	############################################################
	# Build equations ##########################################
	############################################################

	eqns = Array{Equation}(undef, 3 * n + 1)
	count = 1

    # specification as function or intermediate variabel does not matter. I could also make this an
    # actual @variable, then I could retrieve it later.
    rR = 5.5e6 .* exp.(.-5680 ./(273 .+ Tsol)) .* (1 .- exp.(.-cR./50))
    q = 1e4 * (Tgas - Tsol)

    # gas energy balance

    eqns[count] = 0 ~  mflux * 1000 * (Tin - Tgas[1]) - q[1] * dx
    count += 1
	for i in 2:n
	        eqns[count] = 0 ~ mflux * 1000 * (Tgas[i-1] - Tgas[i]) - q[i] * dx
		count += 1
	end

	# solids energy balance
	for i in 1:n
        	eqns[count] = D(Tsol[i]) ~ q[i] / (1000 * 2000)
		count += 1
	end

	# gas momentum balance
	eqns[count] = 0 ~ - mflux[1] + 5e-3 * sqrt(dp)
	count += 1

	# reactant concentration equations
	for i in 1:n
        	eqns[count] = D(cR[i]) ~ -rR[i]
		count += 1
	end
    @named sys = ODESystem(eqns, t, vcat(Tgas, Tsol, cR, [mflux]), params)
    return sys

end


dp = 100e2
Tin = 200
h = 0.5
parameters = [dp, Tin, h]

N = 100
tspan = (0.0, 1500.0)
sys = build_system(N)

# quick and dirty definition of u0, should be Pairs instead
u0 = ones(length(sys.states))
u0[1:N] .= parameters[2]
u0[N+1 : 2*N] .= 20
u0[2*N+1 : 3*N] .= 3000
u0[3*N+1] = 0.5

du0 = 0 .* copy(u0)
prob3 = DAEProblem(sys, du0, u0, tspan, parameters, jac=true, sparse=true)
prob3.f.jac_prototype == nothing
prob3.f.jac == nothing
```

We might want to make this throw a warning until it's handled. 

https://discourse.julialang.org/t/modelingtoolkit-building-a-daeproblem-and-solving-it-with-ida-with-klu-jacobian-usage/77814

### PR DESCRIPTION
**Pull request by Abhishek-1Bhatt**
Aims to add `jac` and `sparse` implementation for `DAEProblem` and resolve #1482

### CODE REVIEW

File: src/systems/diffeqs/abstractodesystem.jl [hunk 1]
```diff
@@ -18,15 +18,15 @@ function calculate_tgrad(sys::AbstractODESystem;
...
+                            sparse=false, simplify=false, dvs=states(sys))
     cache = get_jac(sys)[]
     if cache isa Tuple && cache[2] == (sparse, simplify)
         return cache[1]
     end
     rhs = [eq.rhs for eq ∈ full_equations(sys)]
 
     iv = get_iv(sys)
-    dvs = states(sys)
+    # dvs = states(sys)
```
Comment (ChrisRackauckas) [inline src/systems/diffeqs/abstractodesystem.jl:29]:
```suggestion
```

File: src/systems/diffeqs/abstractodesystem.jl [hunk 2]
```diff
@@ -427,31 +436,49 @@ respectively.
...
+        tgrad_gen = generate_tgrad(sys, dvs, ps;
+                                   simplify=simplify,
+                                   expression=Val{eval_expression}, expression_module=eval_module,
+                                   checkbounds=checkbounds, kwargs...)
+        tgrad_oop, tgrad_iip = eval_expression ? (@RuntimeGeneratedFunction(eval_module, ex) for ex in tgrad_gen) : tgrad_gen
+        _tgrad(u,p,t) = tgrad_oop(u,p,t)
+        _tgrad(J,u,p,t) = tgrad_iip(J,u,p,t)
+    else
+        _tgrad = nothing
+    end
```
Comment (ChrisRackauckas) [inline src/systems/diffeqs/abstractodesystem.jl:462]: DAEFunction doesn't have a tgrad IIRC. Double check.

**Comment by ChrisRackauckas** [comment]
This looks right to me. Add a test?

**Comment by Abhishek-1Bhatt** [comment]
Trying to get this test case(Robertson equations) to work, but `ˍ₋gamma` is casusing trouble

```julia
cd(@__DIR__)
using Pkg
Pkg.activate("ModelingToolkit")

using ModelingToolkit
using Sundials
using Plots
using BenchmarkTools

@parameters k1 k2 k3
@variables t u1(t) u2(t) u3(t)

D = Differential(t)

eqs = [D(u1) ~ -k1*u1 + k3*u2*u3,
       D(u2) ~ k1*u1 - k3*u2*u3 - k2*u2^2,
       0 ~ u1 + u2 + u3 - 1]

@named sys = ODESystem(eqs)

u0 = [u1 => 1.0,
      u2 => 0.0,
      u3 => 0.0]

p = [k1 => 0.04,
     k2 => 3e7,
     k3 => 1e4]

tspan = (0.0, 1e5)

du0 = [-0.04, 0.04, 0.0]

prob = DAEProblem(sys, du0, u0, tspan, p, jac=true, sparse=true)

sol = solve(prob, IDA())

plot(sol, xscale=:log10, tspan=(1e-6, 1e5), layout=(3,1))

@benchmark solve(prob, IDA())
```

File: src/systems/diffeqs/abstractodesystem.jl [hunk 3]
```diff
@@ -428,30 +436,34 @@ function DiffEqBase.DAEFunction{iip}(sys::AbstractODESystem, dvs = states(sys),
...
-    #=
-        # TODO: We don't have enought information to reconstruct arbitrary state
-    =#
+    if jac
+        jac_gen = generate_dae_jacobian(sys, dvs, ps;
+                                        simplify=simplify, sparse=sparse,
+                                        expression=Val{eval_expression}, expression_module=eval_module,
+                                        checkbounds=checkbounds, kwargs...)
+        jac_oop, jac_iip = eval_expression ? (@RuntimeGeneratedFunction(eval_module, ex) for ex in jac_gen) : jac_gen
+        _jac(du,u,p,t) = jac_oop(du,u,p,t)
```
Comment (ChrisRackauckas) [inline src/systems/diffeqs/abstractodesystem.jl:457]:
```suggestion
        _jac(u,p,ˍ₋gamma,t) = jac_oop(u,p,ˍ₋gamma,t)

```

File: src/systems/diffeqs/abstractodesystem.jl [hunk 4]
```diff
@@ -78,6 +77,15 @@ function generate_control_jacobian(sys::AbstractODESystem, dvs = states(sys), ps
     return build_function(jac, dvs, ps, get_iv(sys); kwargs...)
 end
 
+function generate_dae_jacobian(sys::AbstractODESystem, dvs = states(sys), ps = parameters(sys); simplify=false, sparse=false, kwargs...)
+    jac_u = calculate_jacobian(sys; simplify=simplify, sparse=sparse)
+    jac_du = calculate_jacobian(sys; simplify=simplify, sparse=sparse, dvs=Differential(get_iv(sys)).(states(sys)))
+    dvs = states(sys)
+    @variables ˍ₋gamma
+    jac = ˍ₋gamma*jac_du + jac_u
+    return build_function(jac, dvs, ps, get_iv(sys); kwargs...)
```
Comment (ChrisRackauckas) [inline src/systems/diffeqs/abstractodesystem.jl:86]:
```suggestion
    return build_function(jac, dvs, ps, ˍ₋gamma, get_iv(sys); kwargs...)

```

Comment (ChrisRackauckas) [reply src/systems/diffeqs/abstractodesystem.jl:86]: This makes the function `f(du,u,p,gamma,t)`.

**Comment by ChrisRackauckas** [comment]
wow VS Code is absolutely garbage. Okay, so I updated quite a bit in here. The example isn't quite working yet but it's really close. I'm going to merge so we can get good diffs again, and let's try this from here. What I would recommend to debug it would be, write down the analytical solution to the Jacobian and call `f.jac` directly. See if you are getting the right numbers. Slowly build up the tests.

For tests of using Sundials like this, see https://github.com/SciML/Sundials.jl/blob/master/test/common_interface/jacobians.jl.

You were missing the `KLU` part:

```julia
using ModelingToolkit
using Sundials
using Plots
using BenchmarkTools

@parameters k1 k2 k3
@variables t u1(t) u2(t) u3(t)

D = Differential(t)

eqs = [D(u1) ~ -k1 * u1 + k3 * u2 * u3,
    D(u2) ~ k1 * u1 - k3 * u2 * u3 - k2 * u2^2,
    0 ~ u1 + u2 + u3 - 1]

@named sys = ODESystem(eqs)

u0 = [u1 => 1.0,
    u2 => 0.0,
    u3 => 0.0]

p = [k1 => 0.04,
    k2 => 3e7,
    k3 => 1e4]

tspan = (0.0, 1e5)

du0 = [-0.04, 0.04, 0.0]

prob = DAEProblem(sys, du0, u0, tspan, p, jac=true, sparse=true)

sol = solve(prob, IDA(linear_solver=:KLU))

plot(sol, xscale=:log10, tspan=(1e-6, 1e5), layout=(3, 1))

@benchmark solve(prob, IDA())
```

**Comment by isaacsas** [comment]
Out of curiosity, what did VSCode do? (Would like to watch out for it myself in the future…)

**Comment by ChrisRackauckas** [comment]
It reformatted the whole file by changing all whitespacing.

**Pull request merged**